doc/oomd/SONiC OOM daemon.md

```diff
@@ -0,0 +1,255 @@
+
+%% OOMD subscript memory utilzation event from cgroup
+CGROUP(cgroup) -. memory utilzation event .-> OOMD; 
+
+%% OOMD get user info from /etc/passwd
+passwd(/etc/passwd) -. user information .-> OOMD;
+
+%% OOMD terminate user session/procress when memory utilzation reach high-water mark
+OOMD -- terminate signal --> process[process]
+
+
+```
+
+## 3.1 OOMD Implementation
+ - OOMD will be a service managed by systemd.
+ - Get system memory information from /proc/meminfo.
+ - Subscript cgroup memory utilization event from /sys/fs/cgroup/memory/memory.usage_in_bytes.
+ - Get user information from /etc/passwd.
+ - OOMD will send SIGKILL/SIGTERM to kill/terminate process.
+ - OOMD can config to terminate user session, this is because when swap not enabled, Linux system may enters a livelocked state much more quickly and may prevent oomd from responding in a reasonable amount of time. 
+ - The following diagram show how OOMD work:
+    - UserInfoProvider will get user information from /etc/passwd
+    - ConfigProvider will subscript config DB change event and update OOMD config.
+    - MemMonitor will monitor cgroup event and trigger OomHandler.
+    - When memory utilization reach high-water mark, OomHandler will terminate user process to prevent OOM happen. 
+
+```mermaid
+graph TD;
+%% USER information provider response for get user information from /etc/passwd
+passwd(/etc/passwd) -. read user info .-> userinfoprovider[UserInfoProvider];
+userinfoprovider -- update --> oomhandler[OomHandler];
+
+%% Memory monitor response for minitor memory utilzation
+cgroup(cgroup) -. subscribe event .-> memmonitor[MemMonitor]
+memmonitor -- reach high-water mark --> oomhandler[OomHandler]
+
+%% Config provider response for monitor and get latest OOMD config
+confdb[(Config DB)] -. config change event .-> confprovider[ConfigProvider]
+confprovider --> memmonitor
+confprovider --> oomhandler
+
+%% OOM handler will terminate process
+oomhandler -- SIG_TERM--> process[process]
+
+```
+### Other solution for user space OOMD
+
+|                      | Cons                                                         | URL                                                          |
+| -------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| earlyoom             | Designed to free swap, SONiC not have swap.                  | [rfjakob/earlyoom: earlyoom - Early OOM Daemon for Linux (github.com)](https://github.com/rfjakob/earlyoom) |
+| oomd                 | Developed by Facebook, Highly customizable by support plugin. Too heavy for SONiC. | [facebookincubator/oomd: A userspace out-of-memory killer (github.com)](https://github.com/facebookincubator/oomd) |
+| systemd-oomd.service | Require cgroupv2, not scalable, can't customize to only termi. | [systemd-oomd.service(8) - Linux manual page (man7.org)](https://man7.org/linux/man-pages/man8/systemd-oomd.service.8.html) |
```

Comment:
>not scalable [](http://example.com/codeflow?start=43&length=12)

Why cgroup v2 is not scalable?

Reply:
Hua: improve this description.

Reply:
Fixed, systemd-oomd.service not scalable because can't customize to protect system process.

Reply:
You already mentioned it, so you can remove "Not scalable"

Reply:
Removed.